step2

Sehun-Kim opened the pull request:
### Step2

1. 게임의 `Player` 역할을 인터페이스로 만듦
- 로그인 할 경우, 로그인한 User에게 `Player`역할을 준다.
- 사용자의 게임 상대가 될 수 있는 `AutoComPlayer`
2. 게임을 실행할 `IndianPokerController`, 게임의 한 턴을 담당할 `ApiTurnController`
- 게임마다 User(`HumanPlayer`), Computer(`AutoComPlayer`) 정보를 세션에 저장해 둔다. (게임이 종료 되면 `AutoComPlayer`의 정보는 세션에서 제거할 것.)
- Controller의 메소드마다 접근권한과 상태를 가져오기 위해 @LoginPlayer와 @AutoPlayer 어노테이션을 만들고, `HandlerMethodArgumentResolver`를 통해 매개변수로 접근가능하게 하였음.
- 한 게임(IndianPoker)내의 한 턴(`TurnController`)에서 턴 중재 역할을 할 `Dealer`를 Bean으로 등록(session scope)

Commit:
step2

1. @LoginPlayer, @AutoPlayer 어노테이션, handlerMethodArgumentResolver
2. Turn Controller에서 게임의 한턴을 중재할 Dealer Bean 등록(session scope)

javajigi commented on src/main/java/indianpoker/domain/player/AutoComPlayer.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public Chips showChips() {
+        return null;
```
null 대신 NullObject를 만들어 반환하거나 Optional을 쓰는 것은 어떨까?

javajigi commented on src/main/java/indianpoker/domain/player/Player.java:
```diff
@@ -0,0 +1,9 @@
+import support.domain.Deck;
+
+public interface Player {
+    Player initPlayer(Deck deck, Chips chips, boolean firstBetter);
```
메소드가 아니라 추상 클래스를 만들고 생성자를 통해 강제하는 것은 어떨까?

javajigi commented on src/main/java/indianpoker/vo/Chips.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public Chips giveChips(int numberOfMinusChips) {
+        if (numberOfChips < numberOfMinusChips) throw new IllegalArgumentException("남은 Chips가 부족합니다.");
```
이런 예외 처리를 생성자에서 하는 것은 어떨까?

javajigi commented on src/main/java/indianpoker/web/IndianPokerController.java:
```diff
@@ -1,16 +1,39 @@
+    @PostMapping("/start")
+    public String start(@AutoPlayer Player autoPlayer, @LoginPlayer Player loginPlayer, int firstBetter, int chipsNum) {
+        playerService.initPlayers(loginPlayer, autoPlayer, firstBetter, chipsNum);
+        return "redirect:/api/turn/1/init";
```
1과 같이 하드코딩해도 괜찮은가?

javajigi commented on src/main/java/support/domain/Deck.java:
```diff
@@ -0,0 +1,42 @@
+    private List<Card> deck;
+
+    public Deck() {
+        this.deck = this.shuffle(this.init());
```
Deck을 이와 같이 초기화하면 테스트하기 힘들지 않을까?
Deck을 생성하는 부분과 로직 구현부를 분리하는 것은 어떨까?

javajigi commented on src/main/java/indianpoker/service/PlayerService.java:
```diff
@@ -0,0 +1,25 @@
+public class PlayerService {
+    private static final Logger logger = LoggerFactory.getLogger(PlayerService.class);
+
+    public void initPlayers(Player loginPlayer, Player autoPlayer, int firstBetter, int chipsNum) {
```
메소드 인자가 4개인데 더 줄일 수 있는 방법은 없을까?

javajigi reviewed (changes requested):
구현 잘 했네요. 👍 
피드백 남겼어요.

Commit:
step2 feedback reflect

1. annotation 모두 제거
2. 한 세션별로 유지되어야 할 정보를 Turn class로 만듦
3. Turn은 Session Bean으로 등록
